src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,273 @@
+
+                    // Return actual value (potentially overridden by offsetSet)
+                    return $this->offsetGet($offset);
+                }
+            )
+        );
+    }
+
+    public function jsonSerialize(): array
+    {
+        return iterator_to_array($this);
+    }
+
+    /** @param mixed $offset */
+    public function offsetExists($offset): bool
+    {
+        $offset = (int) $offset;
+        $this->ensureOffsetRead($offset);
+
+        return ! isset($this->unset[$offset]) && ! isset($this->notFound[$offset]);
+    }
+
+    /**
+     * @param mixed $offset
+     * @return mixed
+     */
+    #[ReturnTypeWillChange]
+    public function offsetGet($offset)
+    {
+        $offset = (int) $offset;
+        $this->ensureOffsetRead($offset);
+
+        if (isset($this->unset[$offset]) || isset($this->notFound[$offset])) {
+            trigger_error(sprintf('Undefined offset: %d', $offset), E_USER_WARNING);
+        }
+
+        return array_key_exists($offset, $this->set) ? $this->set[$offset] : $this->read[$offset];
+    }
+
+    /**
+     * @param mixed $offset
+     * @param mixed $value
+     */
+    public function offsetSet($offset, $value): void
+    {
+        if ($offset === null) {
+            $this->readEntirePackedArray();
+
+            $offset = max(array_merge(
+                array_keys($this->read),
+                array_keys($this->set),
+            )) + 1;
+        } else {
+            $offset = (int) $offset;
+        }
+
+        $this->set[$offset] = $value;
+        unset($this->unset[$offset]);
+        unset($this->notFound[$offset]);
+    }
+
+    /** @param mixed $offset */
+    public function offsetUnset($offset): void
+    {
+        $offset = (int) $offset;
+        $this->unset[$offset] = true;
+        unset($this->set[$offset]);
+    }
+
+    /**
+     * @param mixed $value
+     * @return mixed
+     */
+    private function decodeBSONValue($value)
+    {
+        return $this->getLibrary()->canDecode($value)
+            ? $this->getLibrary()->decode($value)
+            : $value;
+    }
+
+    private function ensureOffsetRead(int $offset): void
+    {
+        if (isset($this->set[$offset]) || isset($this->notFound[$offset]) || array_key_exists($offset, $this->read)) {
+            return;
+        }
+
+        if (! $this->bson->has($offset)) {
+            $this->notFound[$offset] = true;
+
+            return;
+        }
+
+        $value = $this->bson->get($offset);
+
+        // Decode value using the codec library if a codec exists
+        $this->read[$offset] = $this->decodeBSONValue($value);
+    }
+
+    private function getLibrary(): CodecLibrary
+    {
+        return $this->library ?? $this->library = new CodecLibrary(new LazyBSONDocumentCodec(), new LazyBSONArrayCodec());
+    }
+
+    private function readEntirePackedArray(): void
+    {
+        if ($this->entirePackedArrayRead) {
+            return;
+        }
+
+        $this->entirePackedArrayRead = true;
```

Comment:
Would it be safer to defer this until _after_ iteration? I'm considering some edge case where decoding BSON fails (e.g. during a call to `offsetSet()`) but the user catches and ignores the exception. That might leave LazyBSONArray in an inconsistent state.